pkg/sql/colexec/execgen/template.go

```diff
@@ -0,0 +1,351 @@
+// every combination of possible values for the type of each template argument.
+//
+// For example, given a template function:
+//
+// // execgen:template<b>
+// func foo (a int, b bool) {
+//   if b {
+//     return a
+//   } else {
+//     return a + 1
+//   }
+// }
+//
+// This function will add 2 new func decls to the AST:
+//
+// func foo_true(a int) {
+//   return a
+// }
+//
+// func foo_false(a int) {
+//   return a + 1
+// }
+func createTemplateFuncVariants(f *dst.File) map[string]*funcInfo {
+	ret := make(map[string]*funcInfo)
+
+	dstutil.Apply(f, func(cursor *dstutil.Cursor) bool {
+		n := cursor.Node()
+		switch n := n.(type) {
+		case *dst.FuncDecl:
+			var templateVars []string
+			var templateDecPosition int
+			for i, dec := range n.Decorations().Start.All() {
+				if matches := templateRe.FindStringSubmatch(dec); matches != nil {
+					match := matches[1]
+					// Match now looks like foo, bar
+					templateVars = strings.Split(match, ",")
+					for i, v := range templateVars {
+						templateVars[i] = strings.TrimSpace(v)
+					}
+					templateDecPosition = i
+					break
+				}
+			}
+			if templateVars == nil {
+				return false
+			}
+			// Remove the template decoration.
+			n.Decs.Start = append(
+				n.Decs.Start[:templateDecPosition],
+				n.Decs.Start[templateDecPosition+1:]...)
+
+			// Process template funcs: find template params from runtime definition
+			// and save in funcInfo.
+			info := &funcInfo{}
+			for _, v := range templateVars {
+				var found bool
+				for i, f := range n.Type.Params.List {
+					// We can safely 0-index here because fields always have at least
+					// one name, and we've already banned the case where they have more
+					// than one. (e.g. func a (a int, b int, c, d int))
+					if f.Names[0].Name == v {
+						if ident, ok := f.Type.(*dst.Ident); !ok || ident.Name != "bool" {
+							panic("can't currently handle non-boolean template variables :)")
+						}
+						info.templateParams = append(info.templateParams, templateParamInfo{
+							fieldOrdinal: i,
+							field:        dst.Clone(f).(*dst.Field),
+						})
+						found = true
+						break
+					}
+				}
+				if !found {
+					panic(fmt.Errorf("template var %s not found", v))
+				}
+			}
+			info.decl = n
+
+			// Delete template params from runtime definition.
+			newParamList := make([]*dst.Field, 0, len(n.Type.Params.List)-len(info.templateParams))
+			for i, field := range n.Type.Params.List {
+				var skip bool
+				for _, p := range info.templateParams {
+					if i == p.fieldOrdinal {
+						skip = true
+						break
+					}
+				}
+				if !skip {
+					newParamList = append(newParamList, field)
+				}
+			}
+			n.Type.Params.List = newParamList
+
+			// Now, make variants for every possible combination of allowed values of
+			// the template variables.
+			argsPossibilities := generateAllTemplateArgs(info.templateParams)
+
+			funcDecs := info.decl.Decs
+			cursor.InsertBefore(&dst.GenDecl{
```

Comment:
Could you add a comment here that this is replacing the template func base with a constant? Also, I think this is in conflict with a TODO comment you've left in the inliner.